File: C++/Hard/WrongAnswer/cherry-pickup_9.cpp
class Solution {
public:
    int cherryPickup(vector<vector<int>>& grid) {
        
        int n = grid.size();
        vector<vector<int>> dp(n, vector<int>(n, 0));
        dp[0][0] = grid[0][0];
        for (int t = 1; t < 2 * n - 1; ++t) {
            vector<vector<int>> dp2(n, vector<int>(n, 0));
            for (int i = max(0, t - n + 1); i <= min(n - 1, t); ++i) {
                for (int p = max(0, t - n + 1); p <= min(n - 1, t); ++p) {
                    int j = t - i, q = t - p;
                    if (grid[i][j] == -1 || grid[p][q] == -1) {
                        dp2[i][p] = -1;
                        continue;
                    }
                    if (i > 0) dp2[i][p] = max(dp2[i][p], dp[i - 1][p]);
                    if (p > 0) dp2[i][p] = max(dp2[i][p], dp[i][p - 1]);
                    if (i > 0 && p > 0) dp2[i][p] = max(dp2[i][p], dp[i - 1][p - 1]);
                    if (dp2[i][p] >= 0) dp2[i][p] += grid[i][j];
                    if (i != p) dp2[i][p] += grid[p][q];
                }
            }
            dp = dp2;
        }
        return max(dp[n - 1][n - 1], 0);
    }
};

File: Java/Hard/RuntimeError/cherry-pickup_9.java
class Solution {
    public int cherryPickup(int[][] grid) {
        
        int n = grid.length;
        int[][][] dp = new int[n][n][n];
        for (int[][] row: dp)
            for (int[] layer: row)
                Arrays.fill(layer, Integer.MIN_VALUE);
        dp[0][0][0] = grid[0][0];
        for (int t = 1; t <= 2*n-2; ++t) {
            for (int i = Math.max(0, t-(n-1)); i <= Math.min(n-1, t); ++i) {
                for (int j = Math.max(0, t-(n-1)); j <= Math.min(n-1, t); ++j) {
                    if (grid[i][t-i] == -1 || grid[j][t-j] == -1) {
                        dp[i][j][t] = -9999999;
                        continue;
                    }
                    int val = grid[i][t-i];
                    if (i != j) val += grid[j][t-j];
                    for (int pi = i-1; pi <= i; ++pi)
                        for (int pj = j-1; pj <= j; ++pj)
                            if (pi >= 0 && pj >= 0)
                                dp[i][j][t] = Math.max(dp[i][j][t], dp[pi][pj][t-1] + val);
                }
            }
        }
        return Math.max(0, dp[n-1][n-1][2*n-2]);
    }
}